File: 112-path-sum/112-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        
        def dfs(root, currentSum):
            if not root:
                return False
            if root:
                currentSum += root.val
                if currentSum == targetSum and not root.left and not root.right:
                    return True
            return dfs(root.left, currentSum) or dfs(root.right, currentSum)
                
        
        return dfs(root, 0)
        

File: 112-path-sum/112-path-sum.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func hasPathSum(root *TreeNode, targetSum int) bool {
    ans := depthFirstSearch(root, 0, targetSum)
    return ans
}

func depthFirstSearch(root *TreeNode, currentSum int, targetSum int) bool {
    if root == nil {
        return false
    }
    currentSum += root.Val
    if root.Left == nil && root.Right == nil && currentSum == targetSum {
        return true
    }
    return depthFirstSearch(root.Left, currentSum, targetSum) || depthFirstSearch(root.Right, currentSum, targetSum)
}